File: frontend/src/physics/reliability.js
/**
 * SIMBES — Motor Confiabilidad y MTBF
 * =====================================
 * @ref Nelson (1982) — Applied Life Data Analysis
 * @ref Meeker & Escobar (1998) — Statistical Methods for Reliability Data
 * @ref API RP 11S1
 */

// ─── Probabilidad de supervivencia ──────────────────────────────

/**
 * R(t) = e^(−t / MTBF)
 * Resultado fundamental: R(MTBF) = e⁻¹ ≈ 36.77%
 *
 * @param {number} t_days
 * @param {number} MTBF_days
 * @returns {number} R(t) ∈ [0, 1]
 */
export function survivalProb(t_days, MTBF_days) {
  return Math.exp(-t_days / MTBF_days);
}

/**
 * Genera curva de supervivencia para visualización.
 *
 * @param {number} MTBF_days
 * @param {number} [nPoints=200]
 * @returns {{ points: Array<{t, R}>, R_at_MTBF, t_50pct, t_10pct }}
 */
export function survivalCurve(MTBF_days, nPoints = 200) {
  const t_max  = 2 * MTBF_days;
  const points = [];
  for (let i = 0; i <= nPoints; i++) {
    const t = (t_max * i) / nPoints;
    points.push({ t, R: survivalProb(t, MTBF_days) });
  }
  return {
    points,
    MTBF_days,
    R_at_MTBF: Math.exp(-1),                          // ≈ 0.3679
    t_50pct:   +(-MTBF_days * Math.log(0.50)).toFixed(1),
    t_10pct:   +(-MTBF_days * Math.log(0.10)).toFixed(1),
  };
}

// ─── MTBF MLE con datos censurados ───────────────────────────────

/**
 * MTBF_MLE = T_total / r
 * T_total = Σ t_fallas + Σ t_censurados
 *
 * Ignorar censurados produce sesgo de sobrevivencia inverso.
 *
 * @param {number[]} failure_times  - Tiempos de falla (días)
 * @param {number[]} censored_times - Tiempos de equipos aún operativos (días)
 * @returns {{ MTBF_days, r_failures, T_total_days, bias_warning }}
 */
export function mtbfMLE(failure_times, censored_times) {
  const r        = failure_times.length;
  const T_fail   = failure_times.reduce((a, b) => a + b, 0);
  const T_cens   = censored_times.reduce((a, b) => a + b, 0);
  const T_total  = T_fail + T_cens;
  const n_total  = r + censored_times.length;

  if (r === 0) return {
    MTBF_days: Infinity, r_failures: 0, n_total,
    T_total_days: T_total, bias_warning: true,
    message: 'Sin fallas. MTBF no estimable. Ampliar período de observación.',
  };

  const MTBF       = T_total / r;
  const bias_warn  = n_total > 0 && (r / n_total) < 0.20;

  return {
    MTBF_days:          +MTBF.toFixed(1),
    r_failures:         r,
    n_total,
    T_total_days:       +T_total.toFixed(1),
    lambda_per_day:     +(1 / MTBF).toFixed(6),
    censoring_fraction: n_total > 0 ? +(censored_times.length / n_total).toFixed(3) : 0,
    bias_warning:       bias_warn,
    message:            `MTBF = ${MTBF.toFixed(0)} días · ${r} fallas de ${n_total} equipos.`
      + (bias_warn ? ' ⚠ < 20% han fallado — usar IC.' : ''),
  };
}

// ─── Intervalos de confianza Chi² ────────────────────────────────

/**
 * Cuantil de la normal estándar — aproximación de Beasley-Springer-Moro.
 * @param {number} p - probabilidad (0–1)
 * @returns {number} z tal que Φ(z) = p
 */
function _normalQuantile(p) {
  if (p <= 0) return -Infinity;
  if (p >= 1) return Infinity;
  const a = [2.515517, 0.802853, 0.010328];
  const b = [1.432788, 0.189269, 0.001308];
  const t = p < 0.5
    ? Math.sqrt(-2 * Math.log(p))
    : Math.sqrt(-2 * Math.log(1 - p));
  const num = a[0] + a[1] * t + a[2] * t * t;
  const den = 1 + b[0] * t + b[1] * t * t + b[2] * t * t * t;
  const z   = t - num / den;
  return p < 0.5 ? -z : z;
}

/**
 * Cuantil de la distribución Chi² — aproximación de Wilson-Hilferty.
 * χ²(p, ν) ≈ ν × (1 − 2/(9ν) + z_p × √(2/(9ν)))³
 *
 * [SIMPLIFIED: precisión ±0.5% en rangos ν ≥ 2, 0.01 ≤ p ≤ 0.99]
 *
 * @param {number} p  - probabilidad (0–1)
 * @param {number} df - grados de libertad
 * @returns {number}
 */
function _chi2Inv(p, df) {
  if (df <= 0) return 0;
  const z = _normalQuantile(p);
  const x = 1 - 2 / (9 * df) + z * Math.sqrt(2 / (9 * df));
  return Math.max(0, df * x * x * x);
}

/**
 * Intervalo de confianza Chi² para el MTBF (distribución exponencial).
 *
 * Fórmulas (Nelson 1982):
 *   MTBF_lower = 2T / χ²(1−α/2, 2r+2)   [cuando r ≥ 1]
 *   MTBF_upper = 2T / χ²(α/2, 2r)
 *
 * Donde T = tiempo total acumulado (fallas + censurados).
 *
 * @param {number} T_total      - tiempo total acumulado (días)
 * @param {number} r_failures   - número de fallas observadas
 * @param {number} [alpha=0.10] - nivel de significancia (0.10 = IC 90%)
 * @returns {{ MTBF_lower, MTBF_upper, confidence_pct, chi2_lower_df, chi2_upper_df }}
 *
 * @ref Nelson, W. (1982). Applied Life Data Analysis. Wiley.
 * @ref Meeker & Escobar (1998). Statistical Methods for Reliability Data.
 */
export function chi2ConfidenceInterval(T_total, r_failures, alpha = 0.10) {
  if (r_failures === 0) {
    return {
      MTBF_lower: 0,
      MTBF_upper: Infinity,
      confidence_pct: Math.round((1 - alpha) * 100),
      message: 'Sin fallas observadas. Límite superior no estimable.',
    };
  }

  const chi2_upper = _chi2Inv(1 - alpha / 2, 2 * r_failures + 2);  // denominador MTBF_lower
  const chi2_lower = _chi2Inv(alpha / 2,     2 * r_failures);       // denominador MTBF_upper

  const MTBF_lower = chi2_upper > 0 ? (2 * T_total) / chi2_upper : 0;
  const MTBF_upper = chi2_lower > 0 ? (2 * T_total) / chi2_lower : Infinity;

  return {
    MTBF_lower:      +MTBF_lower.toFixed(1),
    MTBF_upper:      +MTBF_upper.toFixed(1),
    confidence_pct:  Math.round((1 - alpha) * 100),
    chi2_upper_val:  +chi2_upper.toFixed(2),
    chi2_lower_val:  +chi2_lower.toFixed(2),
    alpha,
  };
}

// ─── Detección de sesgo de sobrevivencia ─────────────────────────

/**
 * Evalúa riesgo de sesgo de sobrevivencia en el análisis.
 *
 * @param {number} r_failures
 * @param {number} n_total
 */
export function survivalBiasCheck(r_failures, n_total) {
  const pct = (r_failures / n_total) * 100;
  let level, recommendation;
  if (pct < 5)       { level = 'ALTO';     recommendation = 'Muy pocas fallas. MTBF sobreestimado. Extender observación.'; }
  else if (pct < 20) { level = 'MODERADO'; recommendation = 'Menos del 20% han fallado. Reportar límite inferior del IC.'; }
  else if (pct < 50) { level = 'BAJO';     recommendation = 'Datos razonables. IC manejables.'; }
  else               { level = 'MÍNIMO';   recommendation = 'Mayoría con falla registrada. Estimación robusta.'; }
  return { failure_rate_pct: +pct.toFixed(1), bias_level: level, recommendation };
}

File: backend/physics/reliability.py
"""
SIMBES — Módulo de Confiabilidad y MTBF
=========================================
Fuentes:
  - Nelson, W.B. (1982). Applied Life Data Analysis. Wiley.
  - Meeker, W.Q. & Escobar, L.A. (1998). Statistical Methods for Reliability Data. Wiley.
  - API RP 11S1: Recommended Practice for Electrical Submersible Pump Teardown Report.
  - lifelines library: https://lifelines.readthedocs.io
"""

import math
from typing import Optional
from scipy.stats import chi2
import numpy as np


# ══════════════════════════════════════════════════════════════════
#  MTBF — ESTIMACIÓN CON DATOS CENSURADOS
# ══════════════════════════════════════════════════════════════════

def mtbf_mle(
    failure_times: list[float],    # tiempos de falla (días)
    censored_times: list[float],   # tiempos de equipos aún operativos (días) — "censurados"
) -> dict:
    """
    Estima el MTBF usando Máxima Verosimilitud (MLE) con datos censurados.
    Asume distribución exponencial (tasa de falla constante — "período de vida útil").

    Para distribución exponencial:
        MTBF_MLE = T_total / r
        donde T_total = Σ t_fallas + Σ t_censurados
              r = número de fallas

    IMPORTANTE: Ignorar los equipos censurados (aún operativos) produce
    "Sesgo de Sobrevivencia Inverso" — subestima el MTBF real.

    Args:
        failure_times  : lista de tiempos hasta la falla (días). Puede ser vacía.
        censored_times : lista de tiempos de equipos aún en operación (días).

    Returns:
        dict con:
          'MTBF_days'          : MTBF estimado (días)
          'r_failures'         : número de fallas
          'n_total'            : total de equipos (fallados + censurados)
          'T_total_days'       : tiempo acumulado total (días)
          'lambda_per_day'     : tasa de falla (1/días)
          'censoring_fraction' : fracción de equipos censurados (0–1)
          'bias_warning'       : True si < 20% han fallado (sesgo potencial)

    Ref: Nelson (1982); Meeker & Escobar (1998).
    """
    r = len(failure_times)
    T_failures  = sum(failure_times)
    T_censored  = sum(censored_times)
    T_total     = T_failures + T_censored
    n_total     = r + len(censored_times)

    if T_total <= 0:
        raise ValueError("El tiempo total acumulado debe ser positivo.")
    if r == 0:
        return {
            "MTBF_days":          float("inf"),
            "r_failures":         0,
            "n_total":            n_total,
            "T_total_days":       round(T_total, 1),
            "lambda_per_day":     0.0,
            "censoring_fraction": 1.0,
            "bias_warning":       True,
            "message":            "Sin fallas registradas. MTBF no estimable con MLE. Aumentar período de observación.",
        }

    MTBF    = T_total / r
    lam     = 1.0 / MTBF
    cens_frac = len(censored_times) / n_total if n_total > 0 else 0.0
    bias_warn = (r / n_total) < 0.20 if n_total > 0 else True

    return {
        "MTBF_days":          round(MTBF, 1),
        "r_failures":         r,
        "n_total":            n_total,
        "T_total_days":       round(T_total, 1),
        "lambda_per_day":     round(lam, 6),
        "censoring_fraction": round(cens_frac, 3),
        "bias_warning":       bias_warn,
        "message":            (
            f"MTBF estimado = {MTBF:.0f} días con {r} fallas de {n_total} equipos. "
            + ("⚠ Posible sesgo: < 20% han fallado. Usar intervalos de confianza." if bias_warn else "")
        ),
    }


# ══════════════════════════════════════════════════════════════════
#  PROBABILIDAD DE SUPERVIVENCIA
# ══════════════════════════════════════════════════════════════════

def survival_prob(
    t_days: float,
    MTBF_days: float,
) -> float:
    """
    Probabilidad de supervivencia bajo distribución exponencial.

    R(t) = e^(−t / MTBF) = e^(−λt)

    Resultado fundamental: R(MTBF) = e⁻¹ ≈ 0.3679 (36.77%)
    → Solo el 36.77% de los equipos alcanza su MTBF nominal.

    Args:
        t_days    : tiempo de evaluación (días)
        MTBF_days : MTBF estimado (días)

    Returns:
        Probabilidad de supervivencia R(t) ∈ [0, 1]

    Ref: Nelson (1982).
    """
    if MTBF_days <= 0:
        raise ValueError("MTBF_days debe ser positivo")
    return math.exp(-t_days / MTBF_days)


def survival_curve(
    MTBF_days: float,
    t_max_days: Optional[float] = None,
    n_points: int = 200,
) -> dict:
    """
    Genera la curva de supervivencia completa para visualización.

    Args:
        MTBF_days  : MTBF estimado (días)
        t_max_days : tiempo máximo del eje (default: 2 × MTBF)
        n_points   : número de puntos en la curva

    Returns:
        dict con listas 't' (días) y 'R' (probabilidad de supervivencia).
        También incluye el punto R(MTBF) = 0.3679.
    """
    if t_max_days is None:
        t_max_days = 2.0 * MTBF_days

    t_values = np.linspace(0, t_max_days, n_points)
    R_values = np.exp(-t_values / MTBF_days)

    return {
        "t":           t_values.tolist(),
        "R":           R_values.tolist(),
        "MTBF_days":   MTBF_days,
        "R_at_MTBF":   round(math.exp(-1), 4),   # 0.3679
        "t_90pct":     round(-MTBF_days * math.log(0.90), 1),  # t donde R=90%
        "t_50pct":     round(-MTBF_days * math.log(0.50), 1),  # t donde R=50%
        "t_10pct":     round(-MTBF_days * math.log(0.10), 1),  # t donde R=10%
    }


# ══════════════════════════════════════════════════════════════════
#  INTERVALOS DE CONFIANZA — CHI-CUADRADO
# ══════════════════════════════════════════════════════════════════

def mtbf_confidence_interval(
    T_total_days: float,  # tiempo total acumulado (días)
    r_failures: int,      # número de fallas observadas
    confidence: float = 0.90,  # nivel de confianza (0–1), default 90%
) -> dict:
    """
    Intervalos de confianza asimétricos para MTBF (distribución exponencial)
    usando la distribución Chi-cuadrado.

    MTBF_lower = 2T / χ²(1−α/2, 2r+2)
    MTBF_upper = 2T / χ²(α/2,   2r)

    Para r = 0 (sin fallas): límite superior es infinito.

    Args:
        T_total_days : tiempo acumulado total incluyendo censurados (días)
        r_failures   : número de fallas observadas
        confidence   : nivel de confianza (ej. 0.90 para 90%)

    Returns:
        dict con:
          'MTBF_lower'    : límite inferior del IC (días)
          'MTBF_upper'    : límite superior del IC (días)
          'MTBF_mle'      : MTBF puntual MLE (días)
          'confidence_pct': nivel de confianza en %

    Ref: Meeker & Escobar (1998), Cap. 7.
    """
    if T_total_days <= 0:
        raise ValueError("T_total_days debe ser positivo")
    if r_failures < 0:
        raise ValueError("r_failures no puede ser negativo")

    alpha = 1.0 - confidence

    # Límite inferior: χ²(1−α/2, 2r+2)
    chi2_upper = chi2.ppf(1 - alpha / 2, df=2 * r_failures + 2)
    MTBF_lower = (2.0 * T_total_days) / chi2_upper

    # Límite superior: χ²(α/2, 2r)
    if r_failures == 0:
        MTBF_upper = float("inf")
    else:
        chi2_lower = chi2.ppf(alpha / 2, df=2 * r_failures)
        MTBF_upper = (2.0 * T_total_days) / chi2_lower if chi2_lower > 0 else float("inf")

    MTBF_mle = T_total_days / r_failures if r_failures > 0 else float("inf")

    return {
        "MTBF_lower":     round(MTBF_lower, 1),
        "MTBF_upper":     round(MTBF_upper, 1) if MTBF_upper != float("inf") else "∞",
        "MTBF_mle":       round(MTBF_mle, 1)   if MTBF_mle   != float("inf") else "∞",
        "confidence_pct": round(confidence * 100, 0),
        "r_failures":     r_failures,
        "T_total_days":   T_total_days,
        "interpretation": (
            f"Con {confidence*100:.0f}% de confianza, el MTBF real está entre "
            f"{MTBF_lower:.0f} y "
            f"{'∞' if MTBF_upper == float('inf') else f'{MTBF_upper:.0f}'} días."
        ),
    }


# ══════════════════════════════════════════════════════════════════
#  DETECCIÓN DE SESGO DE SOBREVIVENCIA
# ══════════════════════════════════════════════════════════════════

def survival_bias_check(
    r_failures: int,
    n_total: int,
) -> dict:
    """
    Evalúa si el análisis tiene riesgo de sesgo de sobrevivencia inverso.

    Si < 20% de la población ha fallado, los datos censurados dominan
    y el MTBF estimado puede estar significativamente sesgado.

    Args:
        r_failures : número de fallas observadas
        n_total    : total de equipos en el estudio

    Returns:
        dict con evaluación del sesgo y recomendaciones.
    """
    if n_total == 0:
        return {"error": "n_total debe ser mayor que 0"}

    failure_rate_pct = (r_failures / n_total) * 100.0

    if failure_rate_pct < 5.0:
        level = "ALTO"
        msg   = "Muy pocas fallas. El MTBF está altamente sobreestimado. Extender el período de observación."
    elif failure_rate_pct < 20.0:
        level = "MODERADO"
        msg   = "Menos del 20% han fallado. Usar intervalos de confianza amplios. Reportar límite inferior."
    elif failure_rate_pct < 50.0:
        level = "BAJO"
        msg   = "Datos razonables para estimación. Los IC son manejables."
    else:
        level = "MÍNIMO"
        msg   = "Mayoría de equipos con falla registrada. Estimación robusta."

    return {
        "failure_rate_pct":  round(failure_rate_pct, 1),
        "bias_level":        level,
        "r_failures":        r_failures,
        "n_total":           n_total,
        "recommendation":    msg,
    }
